docs: Add initial docs on how to use node local storage

rata opened the pull request:
This adds basic documentation on how to use node local storage using
local persistent volumes in filesystem mode.

Other sections are mentioned, although empty for now, on other ways to
use it that we may want to expand later.

pothos commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+# Using node local storage on kubernetes
```
I would capitalize it (except "on")

invidian replied:
Use https://capitalizemytitle.com/ :)

rata replied:
I wouldn't, but if all agree (and looking quickly it seems that is done in another docs), so did it :)

pothos commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+# Using node local storage on kubernetes
+
+This guide shows different ways to use node local on Lokomotive, to achieve near
```
`+` storage

rata replied:
Fixed, thanks!

pothos commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+ * Persistent Volumes and PVC: https://kubernetes.io/docs/concepts/storage/persistent-volumes/
+ * StorageClasses: https://kubernetes.io/docs/concepts/storage/storage-classes/
+
+More detail about specific volumes used and how to use them on lokomotive will
```
capital L (also below)

rata replied:
Thanks! Did a s/lokomotive/Lokomotive (excluding URLs :))

pothos commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+**NOTE**: This has been tested on Packet, some lokomotive changes need to be
+ported for this to work on other platforms.
+
+## Using local persistent volume: filesystem mode
```
I would call it "Local Persistent Volumes" because it's used as name for a concept

rata replied:
Thanks, fixed!

pothos commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+## Using local persistent volume: filesystem mode
+
+This way of using node local storage works by exposing a filesystem mounted in
+the host as a persistent volume to pods.
```
I would write "Persistent Volume". (The Kubernetes docs even write it as "`PersistentVolume` (PV)".)

rata replied:
Makes sense, changed to capital in all the docoumen. Thanks! :)

pothos commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+the host as a persistent volume to pods.
+
+One persistent volume can be bound to only one PVC, so only one pod will be able
+to use one persistent volume in the host. However, see the section(link) bellow
```
I think no "(link)" is needed if it's the section below – or which one is meant?

rata replied:
True, no link. It was a leftover. Thanks :)

pothos commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+to use one persistent volume in the host. However, see the section(link) bellow
+to see how to create more than one persistent volume per node.
+
+The kubernetes documentation on node local volumes using filesystem mode is
```
captial K

rata replied:
Thanks, changed it in the whole doc :)

pothos commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+The kubernetes documentation on node local volumes using filesystem mode is
+here (mixed with block mode):
+
+	https://kubernetes.io/docs/concepts/storage/volumes/#local
```
It would be good to make it clickable when rendered as markdown.

rata replied:
Right! Improved all the links now, thanks!

pothos commented on docs/advanced/node-local-storage.md:
`+` which …

rata replied:
Thanks!

pothos commented on docs/advanced/node-local-storage.md:
"StatefulSet", "stateful set", or "Stateful Set"? Just choose one style, I would tend to "Stateful Set" if we talk about it as concept and not as configuration part.

rata replied:
Good point, changed it all to "Stateful Set". Thanks!

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+# Using node local storage on kubernetes
+
+This guide shows different ways to use node local on Lokomotive, to achieve near
+bare-metal performance inside Kubernetes pods.
```
```suggestion
bare-metal storage performance inside Kubernetes pods.
```

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+# Using node local storage on kubernetes
+
+This guide shows different ways to use node local on Lokomotive, to achieve near
```
```suggestion
This guide shows different ways to use node local storage on Lokomotive, to achieve near
```

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+# Using node local storage on kubernetes
```
```suggestion
# Using Node Local Storage on Kubernetes
```

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+the host as a persistent volume to pods.
+
+One persistent volume can be bound to only one PVC, so only one pod will be able
+to use one persistent volume in the host. However, see the section(link) bellow
```
```suggestion
to use one persistent volume in the host. However, see the section(link) below
```

I think "link" might be a placeholder for anchor to section below. Still needs to be added. Or just remove it and have `"section below".

rata replied:
Removed, it was a placeholder but removing is better. Thanks!

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+ * Persistent Volumes and PVC: https://kubernetes.io/docs/concepts/storage/persistent-volumes/
+ * StorageClasses: https://kubernetes.io/docs/concepts/storage/storage-classes/
+
+More detail about specific volumes used and how to use them on lokomotive will
```
This sentence doesn't make sense to me.

>More details about specific volumes used

used?

```suggestion
More details about specific volumes used and how to use them on Lokomotive will
```

rata replied:
What I wanted to say is that read only the generalities and not everything about every volume possible (like Local, NFS, etc.) and more details about the types of volume used in the guide will follow.

Would something like:

```
More detail about specific types of volumes used and how to use them on Lokomotive will
follow, so no need to look into volume specifics on the first read.
```
make more sense to you?

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+ * StorageClasses: https://kubernetes.io/docs/concepts/storage/storage-classes/
+
+More detail about specific volumes used and how to use them on lokomotive will
+follow, so no need to look into volume specifics on the first read.
```
```suggestion
follow, so there is no need to look into volume specifics on the first read.
```

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+More detail about specific volumes used and how to use them on lokomotive will
+follow, so no need to look into volume specifics on the first read.
+
+**NOTE**: This has been tested on Packet, some lokomotive changes need to be
```
```suggestion
**NOTE**: This has been tested on Packet provider only, some Lokomotive changes may need to be
```

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+follow, so no need to look into volume specifics on the first read.
+
+**NOTE**: This has been tested on Packet, some lokomotive changes need to be
+ported for this to work on other platforms.
```
```suggestion
ported to other platforms for this to work.
```

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+**NOTE**: This has been tested on Packet, some lokomotive changes need to be
+ported for this to work on other platforms.
+
+## Using local persistent volume: filesystem mode
```
How about:
```suggestion
## Local Persistent Volumes in Filesystem Mode
```

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+to use one persistent volume in the host. However, see the section(link) bellow
+to see how to create more than one persistent volume per node.
+
+The kubernetes documentation on node local volumes using filesystem mode is
```
```suggestion
The Kubernetes documentation on Local Persistent Volumes using filesystem mode is
```

invidian commented on docs/advanced/node-local-storage.md:
```suggestion
There are some limitations you might want to have in mind:
```

invidian commented on docs/advanced/node-local-storage.md:
```suggestion
devices on `/mnt/sda` or paths inside `/mnt` at your convenience.
```

invidian commented on docs/advanced/node-local-storage.md:
```suggestion
This is the only automated setup supported in lokomotive at the time of writing.
```

rata replied:
Why change `at the time of this writing` by `at the time of writing`?

invidian replied:
https://english.stackexchange.com/questions/199381/is-at-the-time-of-writing-correct `at the time of writing` is the better version of this phrase as far as I know.

rata replied:
Oh, thanks!

invidian commented on docs/advanced/node-local-storage.md:
```suggestion
```

invidian commented on docs/advanced/node-local-storage.md:
```suggestion
   to use different one right now.
```

invidian commented on docs/advanced/node-local-storage.md:
I think comment about PR is not really needed here.

>uber experience

uber? What does it mean?
```suggestion
else and some useful tips from uber experience
```

rata replied:
Uber the company, sorry I wasn't clear :D

Why do you think is not needed? Because we can reference the blog post or because it does not adds no value knowing that in this document?

invidian replied:
>Uber the company, sorry I wasn't clear :D

Maybe `Uber's experience` then?

>Why do you think is not needed? Because we can reference the blog post or because it does not adds no value knowing that in this document?

I would create an issue instead if you would like to file a PR to Kubernetes docs, so we don't forget about it.

rata replied:
Oh, yes! That was an internal note and I will either create an issue or a PR for that. Is that whan you think was not needed? I obviosly agree, that was a mistake :)

invidian commented on docs/advanced/node-local-storage.md:
```suggestion
## Local Persistent Volumes in Raw Block Device Mode
```

invidian commented on docs/advanced/node-local-storage.md:
It's called `OpenEBS Local PV` upstream: https://docs.openebs.io/docs/next/uglocalpv.html
```suggestion
## Using OpenEBS Local PV
```

invidian commented on docs/advanced/node-local-storage.md:
```suggestion
We might consider other container attached storage (CAS) solutions, like [Rook](https://rook.io/),
```

rata commented:
@pothos @invidian thanks a lot for the detailed review! And the commit suggestions!

I think it is much improved now, PTAL @pothos @invidian :)

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+This way of using node local storage works by exposing a filesystem mounted in
+the host as a Persistent Volume to pods.
+
+One Persistent Volume can be bound to only one PVC, so only one pod will be able
```
Too many `one`s here it seems. How about:
```suggestion
Single Persistent Volume can be bound to only one PVC, so only one pod will be able
```

rata replied:
s/Single/A single/?

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+the host as a Persistent Volume to pods.
+
+One Persistent Volume can be bound to only one PVC, so only one pod will be able
+to use one Persistent Volume in the host. However, see the section below to see
```
Too many `one`s here it seems. How about:
```suggestion
to use this Persistent Volume in the host. However, see the section below to see
```

surajssd replied:
Please link to that sub section?

invidian commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,356 @@
+
+### Example
+
+In this example 2 nodes with Local Persistent Volumes will be createad and used
```
```suggestion
In this example 2 nodes with Local Persistent Volumes will be created and used
```

invidian commented on docs/advanced/node-local-storage.md:
```suggestion
### Automatic Local Persistent Volume Creation and Handling
```

invidian commented on docs/advanced/node-local-storage.md:
```suggestion
### Useful Links
```

surajssd commented on docs/advanced/node-local-storage.md:
nit: in a single code block like this if there is a empty line separating multiple commands and it's output it will be helpful.

rata replied:
Thanks, fixed!

surajssd commented on docs/advanced/node-local-storage.md:
This section deserves to be bold, or heading or something or could go into a FAQ section of its own.

rata replied:
Not sure to create one section just for one "what if" paragraph... Maybe a sub section in the example?

surajssd replied:
sure

rata replied:
Marked bold, thanks!

surajssd commented on docs/advanced/node-local-storage.md:
I am not sure if rest of this document should be here, rather than helping user it is just confusing user to follow lot of other links.

I would rather let this document be only something we support and have tried, tested and trusted.

rata replied:
Okay... But I think that is useful and I would prefer to have this somewhere, where do you suggest? Or you suggest to just drop this?

surajssd commented on docs/advanced/node-local-storage.md:
Like I said earlier in the previous comment this deserves its own document.

rata replied:
I prefer to start with a single document of node local storage, if at a later stage we want to have more docs (1 doc per section, a brief summary on each section and then link to another doc, etc.) we can do when we add more stuff.

 What do you think?

surajssd replied:
I want to have a simple and consice documentation for the users so that we can just point them to do certain task. Like for manual node local PV I would point someone to this doc. The doc should not confuse user.

rata replied:
Can you please elaborate on why would this confuse the user? I only see an almost empty section that says "we will expand on this later, these links might be relevant for that".

Would an empty section or removing this be better from your point of view? I'm not sure what to do to address your comment :-/. An new doc with this links doesn't seem great (it's pretty much empty) and not sure what other ways are to address this.

invidian replied:
@surajssd can you look into the modified version? I think it's fine for now to keep it here until we have something more specific. Then we can just refer here to new document.

rata commented:
@surajssd PTAL. I've addressed the all comments, except the ones I asked you something, and adapted to use `/mnt/node-local-storage` as path, as that was recently changed in master

kosyfrances commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,358 @@
+ * [Persistent Volumes and PVC](https://kubernetes.io/docs/concepts/storage/persistent-volumes/)
+ * [StorageClasses](https://kubernetes.io/docs/concepts/storage/storage-classes/)
+
+More details about specific volumes types used and how to use them on Lokomotive
```
s/volumes types/volume types

kosyfrances commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,358 @@
+
+## Local Persistent Volumes in Filesystem Mode
+
+This way of using node local storage works by exposing a filesystem mounted in
```
s/in/on

kosyfrances commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,358 @@
+A single Persistent Volume can be bound to only one PVC in ReadWriteOnce access
+mode, so only one pod will be able to use one Persistent Volume in the host.
+However, see the section below to see how to create more than one Persistent
+Volume per node.
```
How about rephrasing as:

```
See the section below for information on how to create more than one Persistent
Volume per node.
```

kosyfrances commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,358 @@
+block mode).
+	
+Bear in mind that this section is about the filesystem mode, not the block mode.
+That will be covered on a different section of this document.
```
Do you mean that the block mode will be covered in a different section of this document? If yes, how about rephrasing as:

```
Note that this section is about the filesystem mode, not the block mode.
Information about the block mode will be covered on a different section of this document.
```

kosyfrances commented on docs/advanced/node-local-storage.md:
```diff
@@ -0,0 +1,358 @@
+
+### Example
+
+In this example 2 nodes with Local Persistent Volumes will be created and used
```
Add a comma after `example`.

Change `2` to `two` to be consistent with the remaining part of the sentence.

kosyfrances commented on docs/advanced/node-local-storage.md:
Do you mean to say: 
```
Note that when using the `Delete` .....
```

kosyfrances commented on docs/advanced/node-local-storage.md:
```
so you may want to choose the policy that better fits
```
--->
```
so you may want to choose a policy that better fits
```

kosyfrances commented on docs/advanced/node-local-storage.md:
No you won't see it on the host, although if you delete the pod and a new one gets created, it will still have the same contents of the old pod. 

I did a setup using this PR, but whatever I wrote to the pod on `/usr/test-pod`, did not show up on the host's `/mnt/node-local-storage` :confused: . So, I decided to try with a different directory. I created a `/mnt/data` directory on the host and it had the following permissions by default.
```
core@kosy-worker-worker-0 /mnt $ ls -l
total 12
drwxr-xr-x. 2 root root 4096 Sep 23 09:59 data
drwxr-xr-x. 3 root root 4096 Sep 23 07:33 node-local-storage
```
After creating the pv to use `/mnt/data` instead of `node-local-storage`, I wrote to the pod and saw the new file also in `/mnt/data` and noticed that the permissions on `/mnt/data/` changed by itself to:
```
core@kosy-worker-worker-0 /mnt $ ls -ll
total 12
drwxrwsr-x. 2 root bin  4096 Sep 23 10:15 data
drwxr-xr-x. 3 root root 4096 Sep 23 07:33 node-local-storage
```

rata commented:
Node local storage was moved back to `/mnt` in this PR https://github.com/kinvolk/lokomotive-kubernetes/pull/72 recently merged, so we need to adjust this doc to reflect that

invidian commented:
@rata do you have time to take care of this PR?

rata commented:
@invidian not really, and is kind of outdated now. I'll try to keep an eye on this during next week, but no promises :(

invidian commented:
I'm taking a look into it to get it merged.

rata commented on docs/advanced/node-local-storage.md:
This was not the case before self-hosted kubelet, as the pod didn't saw mounts inside mnt (see https://github.com/kinvolk/lokomotive-kubernetes/issues/73) Not sure if it is now possible with self-hosted kubelet

rata replied:
```suggestion
Lokomotive Kubernetes currently only exposes `/mnt/` to the kubelet, so only that
directory can be used for node local storage.
```

rata replied:
If' say remove the "You can **manually** mount different devices on `/mnt/sda` or paths inside `/mnt` at your convenience" but I'm failing to create a github suggestion for this, not sure why :man_facepalming:

rata commented:
IMHO, it looks fine besides the two open comments (the one I just posted and suraj comment from long ago)

rata commented on docs/advanced/node-local-storage.md:
```suggestion

1. Lokomotive Kubernetes currently only exposes `/mnt/` to the kubelet, so only that
directory can be used for node local storage.
```

rata commented on docs/advanced/node-local-storage.md:
For the same issue (https://github.com/kinvolk/lokomotive-kubernetes/issues/73) that was discovered later, I think this _might_ not be accurate either. I'd remove this, probably. What do you think?

The link (the `[1]`), is also shared in the section below, so we still keep a reference for that useful doc :)

invidian replied:
I guess it would work, if we enable mount propagation for this path, which we don't at the moment. I can remove it then.

invidian replied:
Ok, removed now.

rata commented:
@surajssd do you think it looks good to approve the PR? Can you approve it, then? :)

@invidian can you approve too? (not sure, it's weird, you addressed the last comments at least. But I can't approve as it is my PR :))

@pothos or maybe can you take a look and approve if it seems okay? :)

PS: I'd merge with squash and merge, if we are all ok with this (don't see value in the split commits)

invidian commented:
>@surajssd do you think it looks good to approve the PR? Can you approve it, then? :)

@surajssd already approved it, but it got dismissed by my push. Let me squash the commits then.

rata commented:
Merging on behalf of @invidian ? :joy:

rata commented:
@invidian thanks **a lot** for driving this forward! :)
